File: back/app/static/chess_class.js
class Cell
{
    constructor()
    {
        this.piece = null;
        this.possibleMove = false;
        this.move = false;
        this.enPassant = false;
        this.castling = false; // roque
    }
}

class Pawn {
    constructor (color)
    {
        this.color = color;
        this.src = "/static/srcs/chess/" + color + "pawn.svg";
    }
    
    setPossibleMoves(board, x, y)
    {
        if (this.color == 'white')
        {
            if (y == 6 && board[y-1][x].piece == null && board[y-2][x].piece == null)
                board[y-2][x].possibleMove = true;

            if (board[y-1][x].piece == null)
                board[y-1][x].possibleMove = true;
            if (y - 1 >= 0 && x + 1 < 8 && board[y-1][x+1].piece != null && board[y-1][x+1].piece.color != this.color)
                board[y-1][x+1].possibleMove = true;
            if (y - 1 >= 0 && x - 1 >= 0 && board[y-1][x-1].piece != null && board[y-1][x-1].piece.color != this.color)
                board[y-1][x-1].possibleMove = true;
            if (y - 1 == 2 && x + 1 < 8 && board[y-1][x+1].enPassant)
                board[y-1][x+1].possibleMove = true;
            if (y - 1 == 2 && x - 1 >= 0 && board[y-1][x-1].enPassant)
                board[y-1][x-1].possibleMove = true;
        }
        else if (this.color == 'black')
        {
            if (y == 1 && board[y+1][x].piece == null && board[y+2][x].piece == null)
                board[y+2][x].possibleMove = true;

            if (board[y+1][x].piece == null)
                board[y+1][x].possibleMove = true;
            if (y + 1 < 8 && x + 1 < 8 && board[y+1][x+1].piece != null && board[y+1][x+1].piece.color != this.color)
                board[y+1][x+1].possibleMove = true;
            if (y + 1 < 8 && x - 1 >= 0 && board[y+1][x-1].piece != null && board[y+1][x-1].piece.color != this.color)
                board[y+1][x-1].possibleMove = true;
            if (y + 1 == 5 && x + 1 < 8 && board[y+1][x+1].enPassant)
                board[y+1][x+1].possibleMove = true;
            if (y + 1 == 5 && x - 1 >= 0 && board[y+1][x-1].enPassant)
                board[y+1][x-1].possibleMove = true;
        }
    }
}

class Knight {
    constructor (color)
    {
        this.color = color;
        this.src = "/static/srcs/chess/" + color + "knight.svg";
    }
    
    setPossibleMoves(board, x, y)
    {
        if (x - 1 >= 0 && y - 2 >= 0 && (board[y-2][x-1].piece == null || (board[y-2][x-1].piece && board[y-2][x-1].piece.color != this.color)))
            board[y-2][x-1].possibleMove = true;
        if (x + 1 <= 7 && y - 2 >= 0 && (board[y-2][x+1].piece == null || (board[y-2][x+1].piece && board[y-2][x+1].piece.color != this.color)))
            board[y-2][x+1].possibleMove = true;
        if (x - 2 >= 0 && y - 1 >= 0 && (board[y-1][x-2].piece == null || (board[y-1][x-2].piece && board[y-1][x-2].piece.color != this.color)))
            board[y-1][x-2].possibleMove = true;
        if (x + 2 <= 7 && y - 1 >= 0 && (board[y-1][x+2].piece == null || (board[y-1][x+2].piece && board[y-1][x+2].piece.color != this.color)))
            board[y-1][x+2].possibleMove = true;
        if (x - 2 >= 0 && y + 1 <= 7 && (board[y+1][x-2].piece == null || (board[y+1][x-2].piece && board[y+1][x-2].piece.color != this.color)))
            board[y+1][x-2].possibleMove = true;
        if (x + 2 <= 7 && y + 1 <= 7 && (board[y+1][x+2].piece == null || (board[y+1][x+2].piece && board[y+1][x+2].piece.color != this.color)))
            board[y+1][x+2].possibleMove = true;
        if (x - 1 >= 0 && y + 2 <= 7 && (board[y+2][x-1].piece == null || (board[y+2][x-1].piece && board[y+2][x-1].piece.color != this.color)))
            board[y+2][x-1].possibleMove = true;
        if (x + 1 <= 7 && y + 2 <= 7 && (board[y+2][x+1].piece == null || (board[y+2][x+1].piece && board[y+2][x+1].piece.color != this.color)))
            board[y+2][x+1].possibleMove = true;
    }
}

class Rook {
    constructor (color)
    {
        this.color = color;
        this.src = "/static/srcs/chess/" + color + "rook.svg";
    }
    
    setPossibleMoves(board, x, y)
    {
        for (let i = x + 1; i < 8; i++)
        {
            if (verif_loop_moove(board, this.color, i, y))
                break;  
        }
        for (let i = x - 1; i >= 0; i--)
        {
            if (verif_loop_moove(board, this.color, i, y))
                break;
        }
        for (let j = y + 1; j < 8; j++)
        {
            if (verif_loop_moove(board, this.color, x, j))
                break;
        }
        for (let j = y - 1; j >= 0; j--)
        {
            if (verif_loop_moove(board, this.color, x, j))
                break;
        }
    }
}

class Bishop {
    constructor (color)
    {
        this.color = color;
        this.src = "/static/srcs/chess/" + color + "bishop.svg";
    }
    
    setPossibleMoves(board, x, y)
    {
        for (let i = x + 1, j = y + 1; i < 8 && j < 8; i++, j++)
        {
            if (verif_loop_moove(board, this.color, i, j))
                break;
        }
        for (let i = x - 1, j = y - 1; i >= 0 && j >= 0; i--, j--)
        {
            if (verif_loop_moove(board, this.color, i, j))
                break;
        }
        for (let i = x + 1, j = y - 1; i < 8 && j >= 0; i++, j--)
        {
            if (verif_loop_moove(board, this.color, i, j))
                break;
        }
        for (let i = x - 1, j = y + 1; i >= 0 && j < 8; i--, j++)
        {
            if (verif_loop_moove(board, this.color, i, j))
                break;
        }    
    }
}

class King {
    constructor (color)
    {
        this.color = color;
        this.src = "/static/srcs/chess/" + color + "king.svg";
    }
    
    setPossibleMoves(board, x, y)
    {
        if (x - 1 >= 0 && (board[y][x-1].piece == null || board[y][x-1].piece.color != this.color))
            board[y][x-1].possibleMove = true;
        if (x + 1 <= 7 && (board[y][x+1].piece == null || board[y][x+1].piece.color != this.color))
            board[y][x+1].possibleMove = true;
        if (y - 1 >= 0 && (board[y-1][x].piece == null || board[y-1][x].piece.color != this.color))
            board[y-1][x].possibleMove = true;
        if (y + 1 <= 7 && (board[y+1][x].piece == null || board[y+1][x].piece.color != this.color))
            board[y+1][x].possibleMove = true;
        if (x - 1 >= 0 && y - 1 >= 0 && (board[y-1][x-1].piece == null || board[y-1][x-1].piece.color != this.color))
            board[y-1][x-1].possibleMove = true;
        if (x + 1 <= 7 && y - 1 >= 0 && (board[y-1][x+1].piece == null || board[y-1][x+1].piece.color != this.color))
            board[y-1][x+1].possibleMove = true;
        if (x - 1 >= 0 && y + 1 <= 7 && (board[y+1][x-1].piece == null || board[y+1][x-1].piece.color != this.color))
            board[y+1][x-1].possibleMove = true;
        if (x + 1 <= 7 && y + 1 <= 7 && (board[y+1][x+1].piece == null || board[y+1][x+1].piece.color != this.color))
            board[y+1][x+1].possibleMove = true;
        
        // Castling (roque)
        if (!board[y][x].move && !board[y][0].move && board[y][1].piece == null && board[y][2].piece == null && board[y][3].piece == null)
        {
            board[y][2].possibleMove = true;
            board[y][2].castling = true;
        }
        if (!board[y][x].move && !board[y][7].move && board[y][5].piece == null && board[y][6].piece == null)
        {
            board[y][6].possibleMove = true;
            board[y][6].castling = true;
        }
    }
}

class Queen {
    constructor (color)
    {
        this.color = color;
        this.src = "/static/srcs/chess/" + color + "queen.svg";
    }
    
    setPossibleMoves(board, x, y)
    {
        new Rook(this.color).setPossibleMoves(board, x, y);
        new Bishop(this.color).setPossibleMoves(board, x, y);
    }
}


function verif_loop_moove(board, my_color, x, y)
{
    if (board[y][x].piece == null)
    {
        board[y][x].possibleMove = true;
        return 0;
    }
    else if (board[y][x].piece.color != my_color)
    {
        board[y][x].possibleMove = true;
        return 1;
    }
    else
        return 1;
}

File: back/app/consumers/utils/chess_class.py
import json


class Cell:
	def __init__(self):
		self.piece = 0
		self.possibleMove = 0
		self.move = 0
		self.enPassant = 0
		self.castling = 0 # roque


	def toJSON(self):
		mydict = {"piece": self.piece, "possibleMove": self.possibleMove, "move": self.move, "enPassant": self.enPassant, "castling": self.castling}
		return mydict

	def __repr__(self):
		return "{piece: " + str(self.piece) + ", possibleMove: " + str(self.possibleMove) + ", move: " + str(self.move) + ", enPassant: " + str(self.enPassant) + ", castling: " + str(self.castling) + "}"


class Pawn:
	def __init__(self, color):
		self.color = color
	
	def setPossibleMoves(self, board, x, y):
		if (self.color == 'white'):
			if (y == 6 and board[y-1][x]['piece'] == 0 and board[y-2][x]['piece'] == 0):
				board[y-2][x]['possibleMove'] = 1
				board[y-1][x]['enPassant'] = 1

			if (board[y-1][x]['piece'] == 0):
				board[y-1][x]['possibleMove'] = 1
			if (y - 1 >= 0 and x + 1 < 8 and board[y-1][x+1]['piece'] != 0 and board[y-1][x+1]['piece']['color'] != self.color):
				board[y-1][x+1]['possibleMove'] = 1
			if (y - 1 >= 0 and x - 1 >= 0 and board[y-1][x-1]['piece'] != 0 and board[y-1][x-1]['piece']['color'] != self.color):
				board[y-1][x-1]['possibleMove'] = 1
			if (y - 1 == 2 and x + 1 < 8 and board[y-1][x+1]['enPassant']):
				board[y-1][x+1]['possibleMove'] = 1
			if (y - 1 == 2 and x - 1 >= 0 and board[y-1][x-1]['enPassant']):
				board[y-1][x-1]['possibleMove'] = 1
		elif (self.color == 'black'):
			if (y == 1 and board[y+1][x]['piece'] == 0 and board[y+2][x]['piece'] == 0):
				board[y+2][x]['possibleMove'] = 1
				board[y+1][x]['enPassant'] = 1

			if (board[y+1][x]['piece'] == 0):
				board[y+1][x]['possibleMove'] = 1
			if (y + 1 < 8 and x + 1 < 8 and board[y+1][x+1]['piece'] != 0 and board[y+1][x+1]['piece']['color'] != self.color):
				board[y+1][x+1]['possibleMove'] = 1
			if (y + 1 < 8 and x - 1 >= 0 and board[y+1][x-1]['piece'] != 0 and board[y+1][x-1]['piece']['color'] != self.color):
				board[y+1][x-1]['possibleMove'] = 1
			if (y + 1 == 5 and x + 1 < 8 and board[y+1][x+1]['enPassant']):
				board[y+1][x+1]['possibleMove'] = 1
			if (y + 1 == 5 and x - 1 >= 0 and board[y+1][x-1]['enPassant']):
				board[y+1][x-1]['possibleMove'] = 1

	def toJSON(self):
		mydict = {"type": "Pawn", "color": self.color}
		return mydict

	def __repr__(self):
		return "{type: 'Pawn', color: '" + self.color + "'}"


class Rook:
	def __init__(self, color):
		self.color = color
	
	def setPossibleMoves(self, board, x, y):
		for i in range(x + 1, 8):
			if (verif_loop_moove(board, self.color, i, y)):
				break
		for i in range(x - 1, -1, -1):
			if (verif_loop_moove(board, self.color, i, y)):
				break
		for j in range(y + 1, 8):
			if (verif_loop_moove(board, self.color, x, j)):
				break
		for j in range(y - 1, -1, -1):
			if (verif_loop_moove(board, self.color, x, j)):
				break


	def toJSON(self):
		mydict = {"type": "Rook", "color": self.color}
		return mydict

	def __repr__(self):
		return "{type: 'Rook', color: '" + self.color + "'}"


class Knight:
	def __init__(self, color):
		self.color = color
	
	def setPossibleMoves(self, board, x, y):
		if (x - 1 >= 0 and y - 2 >= 0 and (board[y-2][x-1]['piece'] == 0 or (board[y-2][x-1]['piece'] and board[y-2][x-1]['piece']['color'] != self.color))):
			board[y-2][x-1]['possibleMove'] = 1
		if (x + 1 <= 7 and y - 2 >= 0 and (board[y-2][x+1]['piece'] == 0 or (board[y-2][x+1]['piece'] and board[y-2][x+1]['piece']['color'] != self.color))):
			board[y-2][x+1]['possibleMove'] = 1
		if (x - 2 >= 0 and y - 1 >= 0 and (board[y-1][x-2]['piece'] == 0 or (board[y-1][x-2]['piece'] and board[y-1][x-2]['piece']['color'] != self.color))):
			board[y-1][x-2]['possibleMove'] = 1
		if (x + 2 <= 7 and y - 1 >= 0 and (board[y-1][x+2]['piece'] == 0 or (board[y-1][x+2]['piece'] and board[y-1][x+2]['piece']['color'] != self.color))):
			board[y-1][x+2]['possibleMove'] = 1
		if (x - 2 >= 0 and y + 1 <= 7 and (board[y+1][x-2]['piece'] == 0 or (board[y+1][x-2]['piece'] and board[y+1][x-2]['piece']['color'] != self.color))):
			board[y+1][x-2]['possibleMove'] = 1
		if (x + 2 <= 7 and y + 1 <= 7 and (board[y+1][x+2]['piece'] == 0 or (board[y+1][x+2]['piece'] and board[y+1][x+2]['piece']['color'] != self.color))):
			board[y+1][x+2]['possibleMove'] = 1
		if (x - 1 >= 0 and y + 2 <= 7 and (board[y+2][x-1]['piece'] == 0 or (board[y+2][x-1]['piece'] and board[y+2][x-1]['piece']['color'] != self.color))):
			board[y+2][x-1]['possibleMove'] = 1
		if (x + 1 <= 7 and y + 2 <= 7 and (board[y+2][x+1]['piece'] == 0 or (board[y+2][x+1]['piece'] and board[y+2][x+1]['piece']['color'] != self.color))):
			board[y+2][x+1]['possibleMove'] = 1

	def toJSON(self):
		mydict = {"type": "Knight", "color": self.color}
		return mydict

	def __repr__(self):
		return "{type: 'Knight', color: '" + self.color + "'}"



class Bishop:
	def __init__(self, color):
		self.color = color
	
	def setPossibleMoves(self, board, x, y):
		for i,j in zip(range(x + 1, 8), range(y + 1, 8)):
			if (verif_loop_moove(board, self.color, i, j)):
				break
		for i,j in zip(range(x - 1, -1, -1), range(y - 1, -1, -1)):
			if (verif_loop_moove(board, self.color, i, j)):
				break
		for i,j in zip(range(x + 1, 8), range(y - 1, -1, -1)):
			if (verif_loop_moove(board, self.color, i, j)):
				break
		for i,j in zip(range(x - 1, -1, -1), range(y + 1, 8)):
			if (verif_loop_moove(board, self.color, i, j)):
				break

	def toJSON(self):
		mydict = {"type": "Bishop", "color": self.color}
		return mydict

	def __repr__(self):
		return "{type: 'Bishop', color: '" + self.color + "'}"


class Queen:
	def __init__(self, color):
		self.color = color
	
	def setPossibleMoves(self, board, x, y):
		Rook(self.color).setPossibleMoves(board, x, y)
		Bishop(self.color).setPossibleMoves(board, x, y)

	def toJSON(self):
		mydict = {"type": "Queen", "color": self.color}
		return mydict

	def __repr__(self):
		return "{type: 'Queen', color: '" + self.color + "'}"



class King:
	def __init__(self, color):
		self.color = color
	
	def setPossibleMoves(self, board, x, y):
		if (x - 1 >= 0 and (board[y][x-1]['piece'] == 0 or (board[y][x-1]['piece'] and board[y][x-1]['piece']['color'] != self.color))):
			board[y][x-1]['possibleMove'] = 1
		if (x + 1 <= 7 and (board[y][x+1]['piece'] == 0 or (board[y][x+1]['piece'] and board[y][x+1]['piece']['color'] != self.color))):
			board[y][x+1]['possibleMove'] = 1
		if (y - 1 >= 0 and (board[y-1][x]['piece'] == 0 or (board[y-1][x]['piece'] and board[y-1][x]['piece']['color'] != self.color))):
			board[y-1][x]['possibleMove'] = 1
		if (y + 1 <= 7 and (board[y+1][x]['piece'] == 0 or (board[y+1][x]['piece'] and board[y+1][x]['piece']['color'] != self.color))):
			board[y+1][x]['possibleMove'] = 1
		if (x - 1 >= 0 and y - 1 >= 0 and (board[y-1][x-1]['piece'] == 0 or (board[y-1][x-1]['piece'] and board[y-1][x-1]['piece']['color'] != self.color))):
			board[y-1][x-1]['possibleMove'] = 1
		if (x + 1 <= 7 and y - 1 >= 0 and (board[y-1][x+1]['piece'] == 0 or (board[y-1][x+1]['piece'] and board[y-1][x+1]['piece']['color'] != self.color))):
			board[y-1][x+1]['possibleMove'] = 1
		if (x - 1 >= 0 and y + 1 <= 7 and (board[y+1][x-1]['piece'] == 0 or (board[y+1][x-1]['piece'] and board[y+1][x-1]['piece']['color'] != self.color))):
			board[y+1][x-1]['possibleMove'] = 1
		if (x + 1 <= 7 and y + 1 <= 7 and (board[y+1][x+1]['piece'] == 0 or (board[y+1][x+1]['piece'] and board[y+1][x+1]['piece']['color'] != self.color))):
			board[y+1][x+1]['possibleMove'] = 1
		
		# Castling (roque)
		if (not board[y][x]['move'] and not board[y][0]['move'] and board[y][1]['piece'] == 0 and board[y][2]['piece'] == 0 and board[y][3]['piece'] == 0):
			board[y][2]['possibleMove'] = 1
			board[y][2]['castling'] = 1
		if (not board[y][x]['move'] and not board[y][7]['move'] and board[y][5]['piece'] == 0 and board[y][6]['piece'] == 0):
			board[y][6]['possibleMove'] = 1
			board[y][6]['castling'] = 1

	def toJSON(self):
		mydict = {"type": "King", "color": self.color}
		return mydict

	def __repr__(self):
		return "{type: 'King', color: '" + self.color + "'}"








def verif_loop_moove(board, my_color, x, y):
	if (board[y][x]['piece'] == 0):
		board[y][x]['possibleMove'] = 1
		return 0
	elif (board[y][x]['piece']['color'] != my_color):
		board[y][x]['possibleMove'] = 1
		return 1
	else:
		return 1